minifi/minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/command/StartRunner.java

```diff
@@ -0,0 +1,435 @@
+
+        boolean started = waitForStart();
+
+        if (started) {
+            DEFAULT_LOGGER.info("Successfully spawned the thread to start Apache MiNiFi{}", (pid == null ? "" : " with PID " + pid));
+        } else {
+            DEFAULT_LOGGER.error("Apache MiNiFi does not appear to have started");
+        }
+        return process;
+    }
+
+    private boolean needRestart() throws IOException {
+        boolean needRestart = true;
+        File statusFile = bootstrapFileProvider.getStatusFile();
+        if (!statusFile.exists()) {
+            DEFAULT_LOGGER.info("Status File no longer exists. Will not restart MiNiFi");
+            return false;
+        }
+
+        File lockFile = bootstrapFileProvider.getLockFile();
+        if (lockFile.exists()) {
+            DEFAULT_LOGGER.info("A shutdown was initiated. Will not restart MiNiFi");
+            return false;
+        }
+        return needRestart;
+    }
+
+    private void handleReload() {
+        try {
+            Thread.sleep(1000L);
+            if (runMiNiFi.getReloading() && runMiNiFi.isNiFiStarted()) {
+                File swapConfigFile = bootstrapFileProvider.getSwapFile();
+                if (swapConfigFile.exists()) {
+                    DEFAULT_LOGGER.info("MiNiFi has finished reloading successfully and swap file exists. Deleting old configuration.");
+
+                    if (swapConfigFile.delete()) {
+                        DEFAULT_LOGGER.info("Swap file was successfully deleted.");
+                    } else {
+                        DEFAULT_LOGGER.error("Swap file was not deleted. It should be deleted manually.");
+                    }
+                }
+                runMiNiFi.setReloading(false);
+            }
+        } catch (InterruptedException ie) {
+        }
+    }
+
+    private void initConfigFiles(Properties bootstrapProperties, String confDir) throws IOException {
+        File configFile = new File(bootstrapProperties.getProperty(MINIFI_CONFIG_FILE_KEY));
+        try (InputStream inputStream = new FileInputStream(configFile)) {
+            ByteBuffer tempConfigFile = generateConfigFiles(inputStream, confDir, bootstrapProperties);
+            runMiNiFi.getConfigFileReference().set(tempConfigFile.asReadOnlyBuffer());
+        } catch (FileNotFoundException e) {
+            String fileNotFoundMessage = "The config file defined in " + MINIFI_CONFIG_FILE_KEY + " does not exists.";
+            DEFAULT_LOGGER.error(fileNotFoundMessage, e);
+            throw new StartupFailureException(fileNotFoundMessage);
+        } catch (ConfigurationChangeException e) {
+            String malformedConfigFileMessage = "The config file is malformed, unable to start.";
+            DEFAULT_LOGGER.error(malformedConfigFileMessage, e);
+            throw new StartupFailureException(malformedConfigFileMessage);
+        }
+    }
+
+    private Tuple<ProcessBuilder, Process> startMiNiFi() throws IOException {
```

Comment:
This is a fairly large method. It looks like the command building could be refactored out to a separate class, which would make the approach easier to maintain.